### ISSUE TITLE
Importing dynamically written module gives ModuleNotFoundError

### ISSUE THREAD

**Issue by rumkeller** [iodide-project/pyodide#737]
In #419, @mdboom suggests the following way of importing a dynamically created file (without the two commented lines):
```python
import sys
sys.path.insert(0, '.')

# with open("bar.txt", "w") as fd:
#  fd.write("bar")

with open("module.py", "w") as fd:
   fd.write("foo = 42")

import module
module.foo
```

As is, this works fine.

However, with the two lines uncommented this yields `ModuleNotFoundError: No module named 'module'`. This is unexpected, as creating `bar.txt` seems unrelated to the rest of the code.

Adding `import os; print(os.listdir("."))` just before the import shows that both files were indeed created.

Also, with the order of the two `with` blocks swapped it does work as expected.

Tested with v0.15.0 on Firefox 78.0.2.

**Comment by rth** [comment]
Thanks for the report. I can reproduce in Firefox, and it's indeed very puzzling.

**Comment by casatir** [comment]
In [Basthon](https://basthon.fr), users need to import their own modules like this.

With this issue, Pyodide seems limited to import one file per directory. This suggests a workaround : put each module in a separate folder. This is the workaround strategy implemented in Basthon (it should be released soon). Maybe clearer with this code snipet:
```python
import os
import sys

loaded_modules_root = os.path.join("/", "loaded_modules")

for m in ('my_module.py', 'my_other_module.py'):
    root = os.path.join(loaded_modules_root, os.path.splitext(m)[0])
    os.makedirs(root, exist_ok=True)
    with open(os.path.join(root, m), 'w') as f:
        f.write("a = {}\n".format(len(m)))
    sys.path.insert(0, root)

import my_module
print(my_module.a)

import my_other_module
print(my_other_module.a)
```

@rth: does this helps to start investigations?

**Comment by rth** [comment]
Thanks @casatir that's certainly an interesting and puzzling behavior. Wrote a test case in https://github.com/iodide-project/pyodide/commit/dbc0251ccb28ffe20df883c04162a200e497b8a7 (last assert fails there) . What's really puzzling is that `os.listdir`  produces the correct list of files.

It as though there is some per folder cache in the python import system that's not updated when new files are added. We'll need to dig deeper into the Python import system mechanism to understand what is happening.

Though the fact that it happens even with unrelated file types and before the import happens, suggests that either something is not working right at the level of interaction with the filesystem (e.g. some system calls issues in emscripten, though I haven't seen anything related in their issue tracker) or some issue in our in memory compression of the filesystem (https://github.com/iodide-project/pyodide/pull/170). It's oddly specific, though and if it was one of those two I would have expected to catch it when running the CPython test suite..

**Comment by nopid** [comment]
Some caching mechanism seems to be involved indeed. If you call `importlib.invalidate_caches()` before importing, then `bar.txt` does not hide `module` anymore:

```python
import sys
sys.path.insert(0, '.')

with open("bar.txt", "w") as fd:
   fd.write("bar")

with open("module.py", "w") as fd:
   fd.write("foo = 42")

import importlib
importlib.invalidate_caches()
import module
module.foo
```

Tested with v0.15.0 on Firefox 83.0

**Comment by casatir** [comment]
It sounds crystal clear now. Caching mechanism is updated when `bar.txt`is written and not updated afterards.

```python
import sys
sys.path.insert(0, '.')

with open("bar.txt", "w") as fd:
    fd.write("bar")

print(sys.path_importer_cache['.']._path_cache)

with open("module.py", "w") as fd:
    fd.write("foo = 42")

#sys.path_importer_cache['.'].invalidate_caches()

print(sys.path_importer_cache['.']._path_cache)

import module
print(module.foo)
```

Output:
```
{'proc', 'home', 'lib', 'bar.txt', 'tmp', 'dev'}
{'proc', 'home', 'lib', 'bar.txt', 'tmp', 'dev'}
Error: Traceback (most recent call last):
  File "<exec>", line 197, in shell_eval
  File "<exec>", line 165, in eval
  File "/lib/python3.7/site-packages/pyodide.py", line 43, in eval_code
    exec(compile(mod, '<exec>', mode='exec'), ns, ns)
  File "<exec>", line 13, in <module>
ModuleNotFoundError: No module named 'module'
```
Uncommenting the `invalidate_caches` call (which may be lighter than a global `importlib.invalidate_caches()` make it works.

Is it really an issue? Python doc for `invaidate_caches` says

> This function should be called if any modules are created/installed while your program is running to guarantee all finders will notice the new module’s existence.

**Comment by casatir** [comment]
Thought we should leave a comment/warning in #419 in case anyone uses the method presented there.

**Comment by nopid** [comment]
I believe that the problem is with the virtual filesystem. The classical (maybe even POSIX, need to check) behavior for a directory is to change its modification time when a file is created inside the directory. It is not the case in the virtual filesystem. To see it, just run this snippet in a Linux interpreter then inside pyodide console :
```python
from os import stat
stat('.').st_mtime
with open('lapin','w') as f:
  f.write('paf')
stat('.').st_mtime
```

This strange behavior might also impact other libraries.

**Comment by nopid** [comment]
Just a quote to confirm the unexpected behavior of the virtual filesystem and document the expected behavior on file creation according to [POSIX.1-2017](https://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html) :
> If O_CREAT is set and the file did not previously exist, upon successful completion, open() shall mark for update the last data access, last data modification, and last file status change timestamps of the file and the last data modification and last file status change timestamps of the parent directory.

**Comment by dalcde** [comment]
It would make sense to report this to upstream

**Comment by rth** [comment]
Thanks for investigating! Yes, it looks like it's a known emscripten issue https://github.com/emscripten-core/emscripten/issues/11454

**Comment by nopid** [comment]
Without separating timestamps, we can update parent directory timestamp on O_CREAT, it is a one-line change patch, but I have no idea if this would break anything else...

**Comment by nopid** [comment]
tl;dr Let me summarize how timestamps (date and time) work both in Unix-like and in MEMFS (the root virtual filesystem of emscripten) and advocate for the simple change of my last comment.

On a modern Unix-like system, each file (including directories) has 4 timestamps: creation, change, modification and access date and time. Three of them are available through the standard POSIX stat() function call:
```C
    time_t    st_atime;   /* time of last access */
    time_t    st_mtime;   /* time of last modification */
    time_t    st_ctime;   /* time of last status change */
```

In the version of MEMFS provided in [emsdk/emsdk/fastcomp/emscripten/src/library_memfs.js](https://github.com/emscripten-core/emscripten/blob/657ce767f3bb1b0ce5835924df14f70039ba233e/src/library_memfs.js) there is a single attribute per node:
```javascript
      node.timestamp = Date.now();
```

In its sourcecode it is accessed only during the following operations:
 1. [`createNode`](https://github.com/emscripten-core/emscripten/blob/657ce767f3bb1b0ce5835924df14f70039ba233e/src/library_memfs.js#L87) when it is initialized with `Date.now()`
 2. [`getattr`](https://github.com/emscripten-core/emscripten/blob/657ce767f3bb1b0ce5835924df14f70039ba233e/src/library_memfs.js#L174-L176) where the value is read to fill the fields `attr.atime`, `attr.mtime` and `attr.ctime`
 3. [`setattr`](https://github.com/emscripten-core/emscripten/blob/657ce767f3bb1b0ce5835924df14f70039ba233e/src/library_memfs.js#L187-L188) where the value is replaced by the field `attr.timestamp` if filled
 4. [`write`](https://github.com/emscripten-core/emscripten/blob/657ce767f3bb1b0ce5835924df14f70039ba233e/src/library_memfs.js#L293) where it is updated to `Date.now()` each time some data is written to a file (excluding directories and links that do not support the write operation)

Thus the current semantics of `timestamp` in MEMFS is more like a last modification timestamp for files and a creation timestamp for directories.

My proposal is to modify `createNode` to also update the `timestamp` of the parent directory so that the virtual filesystem behaves like the modification timestamp of a [POSIX.1-2017](https://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html) compatible filesystem:

> If O_CREAT is set and the file did not previously exist, upon successful completion, open() shall mark for update the last data access, last data modification, and last file status change timestamps of the file and the last data modification and last file status change timestamps of the parent directory.


The change is very minimal, just add one line to update the parent between [line 89 and line 90 of library_memfs.js](https://github.com/emscripten-core/emscripten/blob/657ce767f3bb1b0ce5835924df14f70039ba233e/src/library_memfs.js#L87-L93). No space overhead and very little time overhead.

Please notice that this is very different from implementing a full POSIX filesystem (space and time overhead) or implementing a separation of access and modification timestamps (as discussed in the upstream issue linked by @rth).

This would fix the caching problems discussed in the current issue. Indeed, Python defines this behavior in `Lib/importlib/_bootstrap_external.py` inside `FileFinder`:
```python
class FileFinder:

    """File-based finder.

    Interactions with the file system are cached for performance, being
    refreshed when the directory the finder is handling has been modified.

    """

(snip)

    def find_spec(self, fullname, target=None):
        """Try to find a spec for the specified module.

        Returns the matching spec, or None if not found.
        """
        is_namespace = False
        tail_module = fullname.rpartition('.')[2]
        try:
            mtime = _path_stat(self.path or _os.getcwd()).st_mtime
        except OSError:
            mtime = -1
        if mtime != self._path_mtime:
            self._fill_cache()
            self._path_mtime = mtime

```

It could of course be discussed upstream but it looks like upstream userbase is more diverse and their proposal would certainly be (see linked PR) to implement this as a different filesystem (that would be mounted inside the MEMFS root fs).

**Comment by nopid** [comment]
...and of course, we have to also update the timestamp of the parent directory on unlink.

**Comment by nopid** [comment]
I now believe that #737 covers 2 differents bugs:
 1. the directory modification time issue that caused import cache not to reload (solved by #893);
 2. something inside pyodide triggers the import cache to be filled too soon leading to file shadowing that modification time cannot always solve.

In fact, #893 solves the bug pointed by @rumkeller only in the case where the test case is executed line by line inside the console. Otherwise, `bar.txt` is written first and pyodide seems to update the directory cache right then. Then `module.py` is written, unfortunately with the same timestamp, thus the directory cache is not updated as the timestamp did not change since last update. This is not the behavior one sees in classical Python: writing files is not supposed to cause cache update, it is only when one imports `module` that the cache is filled (hence with the 2 files inside the cache).

Here is a test that cover both bugs:
```python
exec(compile(r"""
import os
import sys
import tempfile
from pathlib import Path
from time import time

def sleep(duration):
    theend=time()+duration
    while time()<theend:
        pass

tmpdir = Path(tempfile.mkdtemp(dir="/tmp"))

def status():
  incache = str(tmpdir) in sys.path_importer_cache
  print(f'{ tmpdir }: { os.listdir(tmpdir) }')
  print(f'tmpdir in path_importer_cache? { incache }')
  if incache:
    ff = sys.path_importer_cache[str(tmpdir)]
    print(f'  { ", ".join(ff._path_cache) }')

def showtime(fname):
  global tmpdir
  print(f'{ int(os.stat(tmpdir / fname).st_mtime) } { fname }')

def writefile(fname,content):
    global tmpdir
    with open(tmpdir / fname,'w') as f:
        f.write(content)

status()
showtime(tmpdir)
print('\n>>> sys.path.insert')
sys.path.insert(1,str(tmpdir))
status()
showtime(tmpdir)
sleep(1)
print('\n>>> write module_a.py')
writefile('module_a.py','foo = 42\n')
status()
showtime(tmpdir)
showtime(tmpdir / "module_a.py")
print('\n>>> write module_b.py')
writefile('module_b.py','bar = 9*6\n')
status()
showtime(tmpdir)
showtime(tmpdir / "module_a.py")
showtime(tmpdir / "module_b.py")
sleep(1)
print('\n>>> import module_b')
import module_b
assert module_b.bar == 54
status()
showtime(tmpdir)
showtime(tmpdir / "module_a.py")
showtime(tmpdir / "module_b.py")
sleep(1)
print('\n>>> write module_c')
writefile('module_c.py','baz = 17\n')
status()
showtime(tmpdir)
showtime(tmpdir / "module_a.py")
showtime(tmpdir / "module_b.py")
showtime(tmpdir / "module_c.py")
sleep(1)
print('\n>>> import module_c')
import module_c
assert module_c.baz == 17
status()
showtime(tmpdir)
showtime(tmpdir / "module_a.py")
showtime(tmpdir / "module_b.py")
showtime(tmpdir / "module_c.py")
""", "<eval>", mode="exec"), globals())
```

When run inside a classical Python console, on sees that the cache is filled only on import:
```
(...)
>>> write module_b.py
/tmp/tmpq70_m04n: ['module_a.py', 'module_b.py']
tmpdir in path_importer_cache? False
1608712898 /tmp/tmpq70_m04n
1608712898 /tmp/tmpq70_m04n/module_a.py
1608712898 /tmp/tmpq70_m04n/module_b.py

>>> import module_b
/tmp/tmpq70_m04n: ['module_a.py', '__pycache__', 'module_b.py']
tmpdir in path_importer_cache? True
  module_b.py, module_a.py
1608712899 /tmp/tmpq70_m04n
1608712898 /tmp/tmpq70_m04n/module_a.py
1608712898 /tmp/tmpq70_m04n/module_b.py
(...)
```

When run inside the pyodide console, cache is filled too soon:
```
(...)
>>> sys.path.insert
/tmp/tmpazdue_z8: []
tmpdir in path_importer_cache? False
1608713724 /tmp/tmpazdue_z8
>>> write module_a.py
/tmp/tmpazdue_z8: ['module_a.py']
tmpdir in path_importer_cache? True
  module_a.py
1608713724 /tmp/tmpazdue_z8
1608713725 /tmp/tmpazdue_z8/module_a.py
>>> write module_b.py
/tmp/tmpazdue_z8: ['module_a.py', 'module_b.py']
tmpdir in path_importer_cache? True
  module_a.py
1608713724 /tmp/tmpazdue_z8
1608713725 /tmp/tmpazdue_z8/module_a.py
1608713725 /tmp/tmpazdue_z8/module_b.py
(...)
```

Is there something that pyodide does under the hood that would trigger imports of modules in between each executed line? That would explain this behavior.

**Comment by nopid** [comment]
@rth, @dalcde Any clue why writing to a file triggers the update of path_importer_cache? Is there some part of pyodide that watches filesystem access? I digged through the source without success yet.

**Comment by nopid** [comment]
Finally nailed it! It is not a bug, it's a feature :smile: Thanks to `sys.addaudithook` it becomes clear:
```python
exec(compile(r"""
import os
import sys
import tempfile
from pathlib import Path
from time import time

def mook(*args):
    print(">>>> Oook ",args)

sys.addaudithook(mook)

tmpdir = Path(tempfile.mkdtemp(dir="/tmp"))

def status():
  incache = str(tmpdir) in sys.path_importer_cache
  print(f'{ tmpdir }: { os.listdir(tmpdir) }')
  print(f'tmpdir in path_importer_cache? { incache }')
  if incache:
    ff = sys.path_importer_cache[str(tmpdir)]
    print(f'  { ", ".join(ff._path_cache) }')

status()
print('\n>>> sys.path.insert')
sys.path.insert(1,str(tmpdir))
status()
print('\n>>> write module_a.py')
with open(tmpdir / 'module_a.py','w') as f:
    f.write('foo = 42')
status()
""", "<eval>", mode="exec"), globals())
```

writing to a file with specifying the encoding in open triggers the _io module to search for the preferred encoding, loading the `_locale` library.

**Comment by nopid** [comment]
The `_bootlocale` module loaded on the first write in a textfile without encoding is located into `/lib/python3.8/` and `sys.path` contains:
```
/lib/python38.zip,/lib/python3.8,/lib/python3.8/lib-dynload,/lib/python3.8/site-packages
```

Now that the MEMFS modification time issue is fixed, the remaining problems from the current issue comes from inserting a directory inside `sys.path` at position 0 or 1. If you insert in position 2 or append, the bug disappears. Per se, it is not a bug and there is no need to fix it.

**Comment by rth** [comment]
@nopid To summarize should we then close this issue?

**Comment by nopid** [comment]
I believe so. The bug was with the modification timestamp impacting caching. The rest is just expected Python behavior. One shoud either add directory far enough in `sys.path` or open file with specified encoding or import `_bootlocale` first.

**Comment by rth** [comment]
Thanks for working on this @nopid!
